### PR TITLE
Added KV-cache for Nudging

### PR DESCRIPTION
**Pull request by fywalter**
This PR optimizes the Nudging inference class to significantly improve generation speed and makes some slight change to the stopping criteria.

**Key Changes:**

- **KV-Caching Implementation:** Refactored the generation loop to maintain `past_key_values` for both the base and nudging models, reducing complexity from $O(N^2)$ to $O(N)$ per token.
- **Stopping criteria**: Switched from either base or nudging model generate EOS to when the active model (model selected to generate the next token) generates EOS.
- **Documentation:** Updated the `api_nudging.py` docstrings and README to reflect the new `use_kv_cache` parameter and performance trade-offs.

**Performance Impact:**

- Inference on `agieval_tiny` (dev set) with `batch_size=8` is approximately **4x faster** than the previous stateless implementation with `batch_size=1`, and around **10x faster** than without kv-caching with `batch_size=8`.

**Notes for Reviewers:**

- Enabling `use_kv_cache=True` may result in slight output variations due to "greedy locking" of tokenization boundaries and `bfloat16` precision differences between prefill/decode kernels. This is expected behavior for multi-tokenizer collaboration.

### CODE REVIEW

**Comment by BunsenFeng** [comment]
Nice!! Yuyang please test this.

**Comment by leopoldwhite** [comment]
Add config 4-6 and finished successfully

**Review by BunsenFeng — APPROVED** [review]
nice!!

**Comment by BunsenFeng** [comment]
@leopoldwhite please click approve and we are good to go. Yu I also fixed the attention mask thing in `logit_arithmetic`, thank you!

**Review by leopoldwhite — APPROVED** [review]
